### PR TITLE
Fix condition for weights_only unpickler for DTensor

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Same as #140739 but for DTensor (move safe globals for DTensor to `torch.distributed.tensor.__init__` and update error message to let user know `torch.distributed.tensor` must be imported to load DTensor)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140850
* __->__ #140740
* #140739

Differential Revision: [D65961690](https://our.internmc.facebook.com/intern/diff/D65961690)

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65961690).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D65961690).

File: test/distributed/_tensor/test_dtensor.py [hunk 1]
```diff
@@ -542,6 +546,51 @@ def test_dtensor_save_load(self):
...
+            with tempfile.NamedTemporaryFile() as f:
+                torch.save(sharded_tensor, f)
+                import_string = (
+                    "import torch.distributed.tensor" if should_import else ""
+                )
+                script = f"""
+import torch
+{import_string}
+x = torch.load("{f.name}", weights_only=True)
+"""
```
Comment (malfet) [inline test/distributed/_tensor/test_dtensor.py:569]: Nit (no need for 3 line script, use semicolon as line separator. And no need to specify weigh_only as it is a default right now
```suggestion
                script = f"import torch;{import_sring};torch.load('{f.name}')"
```

File: test/distributed/_tensor/test_dtensor.py [hunk 2]
```diff
@@ -542,6 +546,51 @@ def test_dtensor_save_load(self):
...
+                try:
+                    subprocess.check_output(
+                        [sys.executable, "-c", script],
+                        cwd=os.path.dirname(os.path.realpath(__file__)),
+                        stderr=subprocess.STDOUT,
+                    )
+                    if not should_import:
+                        raise RuntimeError(
+                            "Script executed successfully despite lack of import"
+                        )
```
Comment (malfet) [inline test/distributed/_tensor/test_dtensor.py:579]: How about something like that:
```python
cm = self.assertRaisesRegex(RuntimeError, "_picke.Unpickling...") else contextlib.nullcontext()
with cm:
    try:
      subprocess.check_call([sys.executable, "-c", script])
  except subprocess.CalledProcessError as e:
      raise RuntimeError(e.output.decode("utf-8"))
```
This way you can account for case when something should finish without an error and when something should raise an exception

File: test/distributed/_tensor/test_dtensor.py [hunk 3]
```diff
@@ -542,6 +546,51 @@ def test_dtensor_save_load(self):
...
+                )
+                script = f"""
+import torch
+{import_string}
+x = torch.load("{f.name}", weights_only=True)
+"""
+                try:
+                    subprocess.check_output(
+                        [sys.executable, "-c", script],
+                        cwd=os.path.dirname(os.path.realpath(__file__)),
```
Comment (malfet) [inline test/distributed/_tensor/test_dtensor.py:573]: Instead of specifying a working directory, why not path file via absolute path to the script?

**Review by malfet — APPROVED** [review]

**Comment by wz337** [comment]
Thanks for helping to fix it!

File: test/distributed/_tensor/test_dtensor.py (hunk 3 above)
Comment (mikaylagawarecki) [reply test/distributed/_tensor/test_dtensor.py]: For this one actually added this following what was done here, copied over the comment as well

https://github.com/pytorch/pytorch/blob/175ba9fed6a718dd1d0add9c733ed49bab97cc6c/test/test_autograd.py#L10170-L10172

**Pull request closed without merging**